src/validator.cpp

```diff
@@ -723,25 +910,30 @@ void Validator::ValidatorImpl::validateAndCleanMathCiCnNodes(XmlNodePtr &node, c
             }
             attribute = attribute->getNext();
         }
+
+        bool checkUnitsIsInComponent = false;
         // Check that cellml:units has been set.
-        if (!isCellmlIdentifier(unitsName)) {
-            if (nodeType == "cn") {
+        if (nodeType == "ci") {
+            if (unitsAttribute != nullptr) {
                 ErrorPtr err = std::make_shared<Error>();
-                err->setDescription("Math cn element with the value '" + textNode +
-                                    "' does not have a valid cellml:units attribute.");
-                err->setComponent(component);
-                err->setKind(Error::Kind::MATHML);
-                mValidator->addError(err);
-            } else if (node->getParent()->isType("bvar")) {
+                err->setDescription("Math ci element with value '" + textNode +
+                                    "' has a cellml:units attribute with name '" + unitsAttribute->getValue() + "'.");
+            }
+        } else if (nodeType == "cn") {
+            if (isCellmlIdentifier(unitsName)) {
+                checkUnitsIsInComponent = true;
+            } else {
                 ErrorPtr err = std::make_shared<Error>();
-                err->setDescription("Math bvar ci element with the value '" + textNode +
+                err->setDescription("Math cn element with the value '" + textNode +
```

Comment:
Just because I am thinking of it: because we have no information about the location of the error (e.g. line/column in the CellML file), this means that we may end up with several errors with exactly the same description, even for the same (here) component.